registration/views.py

```diff
@@ -56,30 +66,135 @@ def user_data_view(request):
+
+    def get(self, request):
+        """
+        Handle get requests.
+
+        Create a form with the users data.
+        """
+        if not self.profile:
+            return _no_profile_error(request)
 
-    # if verified only ALWAY_WRITABLE fields are writeable
-    if profile.verified:
-        for field in FORM_KEYS:
-            if field not in ALWAYS_WRITABLE:
-                form.fields[field].widget.attrs['readonly'] = True
+        self.initial_data = _get_initial_data(self.user, self.profile)
+        form = UserDataForm(self.initial_data)
 
-    if request.method == 'GET':
         return render(
             request,
-            'registration/user_data.html',
-            {'form': form, 'user': user}
+            self.template_name,
+            {'form': form, 'user': self.user}
         )
 
-    if 'print' in request.POST:
-        return generate_registration_form(user, profile)
-    else:
-        assert 'manual-form' in request.POST
-        return FileResponse(
-            open('files/Nutzerkartei_Anmeldung_2017.pdf', 'rb'),
-            filename=('application_form.pdf')
+    def post(self, request):
+        """
+        Handle post requests.
+
+        Return the registration form as pdf. Either filled with the users data
+        or just as template.
+        """
+        if 'print' in request.POST:
+            assert self.profile
+            return generate_registration_form(self.user, self.profile)
+        else:
+            assert 'manual-form' in request.POST
+            return http.FileResponse(
+                open('files/Nutzerkartei_Anmeldung_2017.pdf', 'rb'),
+                filename=('application_form.pdf')
+            )
+
+
+@method_decorator(login_required, name='setup')
+class EditProfileView(generic.View):
+    """View so the user can edit his/her profile."""
+
+    template_name = 'registration/edit_profile.html'
+
+    def setup(self, request, *args, **kwargs):
+        """Initialize attributes."""
+        self.user, self.profile = _get_user_and_profile(request)
+        self.initial_data = _get_initial_data(self.user, self.profile)
+
+        return super(EditProfileView, self).setup(request, *args, **kwargs)
+
+    def get(self, request):
+        """
+        Handle get requests.
+
+        Create form with user data and writeable fields depending on the users
+        verification.
+        """
+        if not self.profile:
+            return _no_profile_error(request)
+
+        form = UserDataForm(self.initial_data)
+
+        ALWAYS_WRITABLE = ['phone_number', 'mobile_number']
+        # if verified only ALWAY_WRITABLE fields are writeable
+        if self.profile.verified:
+            for field in form.fields.keys():
+                if field not in ALWAYS_WRITABLE:
+                    form.fields[field].widget.attrs['readonly'] = True
+
+        return render(
+            request,
+            self.template_name,
+            {'form': form, 'user': self.user}
         )
```

Comment:
With Django 4.0 the forms have a `render()` [method](https://docs.djangoproject.com/en/4.0/ref/forms/api/#render) on their own. Might it be possible to use that in this case make it easier to read?

Reply:
Using the forms render function, I'm getting trouble with the crispy form package.